packages/core/README.md

```diff
@@ -21,6 +21,40 @@ see binding examples such as
 -   https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-http
 -   https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-coap
 
+## Codecs
+
+### Octet-Stream
+
+The octet-stream codec enables deserialization and serialization of binary data. To encode and decode sequences of bytes, the octet-stream codec uses the following content type parameters:
+| Parameter | Description | Default |
+| --- | --- | --- |
+| `length` | Number of bytes produced during serialization or consumed during deserialization | |
+| `signed` | Signedness of the data type, `true` or `false` | `true` |
+| `byteSeq` | Endianness, enum of `BIG_ENDIAN`, `LITTLE_ENDIAN`, `BIG_ENDIAN_BYTE_SWAP`, `LITTLE_ENDIAN_BYTE_SWAP` | `BIG_ENDIAN` |
+
+Additionally, the octet-stream codec supports the data schema terms below for addressing bit-fields:
+| Term | Description | Default |
+| --- | --- | --- |
+| `ex:bitLength` | Number of bits produced during serialization or consumed during deserialization | |
+| `ex:bitOffset` | Offset in bits from the beginning of the byte sequence | `0` |
+
+With the help of the terms and parameters above, the octet-stream codec can be used to serialize and deserialize objects containing bit-fields and sequences of bytes, like in the following example.
+
+#### Example
+
+To serialize the object `{ flag1: true, numberProperty: 99, stringProperty: "Web" }` to a sequence of bytes, the content type `application/octet-stream;length=4;signed=false;`, along with the following data schema can be used:
+
+```json
+{
+    "type": "object",
+    "properties": {
+        "flag1": { "type": "boolean", "ex:bitOffset": 0, "ex:bitLength": 1 },
+        "numberProperty": { "type": "integer", "ex:bitOffset": 1, "ex:bitLength": 7 },
+        "stringProperty": { "type": "string", "ex:bitOffset": 2, "ex:bitLength": 24 }
```

Comment:
```suggestion
        "stringProperty": { "type": "string", "ex:bitOffset": 8, "ex:bitLength": 24 }
```

I am not very much in the proposal but shouldn't it be 8 instead of 2 .. since 8 bits are used for flag1 and numberProperty.

In general I wonder how one can deal with the order of terms ... since in general a JSON object per se is unordered and hence you need the the right offset, don't you?

Reply:
Good find, thank you!

I'm unsure if this is what you're asking, but the `ex:bitOffset` term describes the serialized octet-stream, so the deserialized JSON object does not need to be ordered.

When deserializing the octet-stream `[0x63, 0x57, 0x65, 0x62]` with the schema defined in the readme, one might as well change the order of the object's properties to

```json
{
    "type": "object",
    "properties": {
        "stringProperty": { "type": "string", "ex:bitOffset": 8, "ex:bitLength": 24 }
        "flag1": { "type": "boolean", "ex:bitOffset": 0, "ex:bitLength": 1 },
        "numberProperty": { "type": "integer", "ex:bitOffset": 1, "ex:bitLength": 7 },
    }
}
```

Also `{ flag1: true, numberProperty: 99, stringProperty: "Web" }`  and `{ numberProperty: 99, flag1: true,  stringProperty: "Web" }`  both get serializied to `[0x63, 0x57, 0x65, 0x62]` with both of the schemas above.

Therefore, the Data Schema specifies the order of the bits in the octet-stream independently from the order in the JSON object.